src/Storages/MergeTree/GinIndexStore.h

```diff
@@ -0,0 +1,267 @@
+#pragma once
+
+#include <array>
+#include <vector>
+#include <unordered_map>
+#include <mutex>
+#include <Core/Block.h>
+#include <Disks/IDisk.h>
+#include <IO/ReadBufferFromFileBase.h>
+#include <IO/WriteBufferFromFileBase.h>
+#include <roaring.hh>
+#include <Common/FST.h>
+#include <Storages/MergeTree/IDataPartStorage.h>
+namespace DB
+{
+
+constexpr int MIN_SIZE_FOR_ROARING_ENCODING = 16;
+
+/// Gin Index Segment information, which contains:
+struct GinIndexSegment
+{
+    /// .gin_post file offset of this segment's postings lists
+    UInt64 postings_start_offset{0};
+
+    /// .gin_dict file offset of this segment's term dictionaries
+    UInt64 item_dict_start_offset{0};
+
+    /// Next row ID for this segment
+    UInt32 next_row_id{1};
+
+    ///  Segment ID retrieved from next available ID from file .gin_sid
+    UInt32 segment_id {0};
+};
+
+using GinIndexSegments = std::vector<GinIndexSegment>;
+
+/// GinIndexPostingsList which uses 32-bit Roaring
+using GinIndexPostingsList = roaring::Roaring;
+
+using GinIndexPostingsListPtr = std::shared_ptr<roaring::Roaring>;
+
+/// Gin Index Postings List Builder.
+struct GinIndexPostingsBuilder
+{
+    /// When the list length is no greater than MIN_SIZE_FOR_ROARING_ENCODING, array 'lst' is used
+    std::array<UInt32, MIN_SIZE_FOR_ROARING_ENCODING> lst;
```

Comment:
How beneficial is that in terms of performance?

Reply:
We found roaring bitmap takes more space than expected for small sets of integers, that's why we use list for small sets.